### ISSUE TITLE
DOC: Controlling (or better documenting) the required/recommended Doxygen version

### ISSUE THREAD

**Issue by eagunn** [numpy/numpy#27610]
### Issue with current documentation:

As of this writing: 
- The CircleCI build uses Doxygen version 1.9.1 [1]
    - Doxygen is installed in the `install system dependencies` step in the `numpy\.circleci\config.yml` file which does not specify either an exact or a minimum version, simply `sudo apt-get install <many other packages> doxygen`
    - I haven't been able to figure out what package repository is used.
    - This build generates one obsolete tag warning.
- Building a new conda numpy-dev environment installs Doxygen 1.10
    - Doxygen is specified in the dependencies section of `numpy\environment.yml`, but neither an exact nor a minimum version is specified
    - The package repository is conda-forge [2] so we simply get whatever version they are doling out today.
    - This version generates several obsolete tag warnings, see #27515.
- Our dev documentation [3]:
   - Implies that the developer needs to install Doxygen 'by hand' when, in fact, it is installed as part of the conda numpy-dev environment.
   - Has a note: `Try to install a newer version of Doxygen > 1.8.10 otherwise you may get some warnings during the build.` 
- The current stable version of Doxygen, if one were to download it from their website, is 1.12 [4]

I recommend that, at a minimum:
- The `Dependencies` section of the dev doc be updated to acknowledge/recommend that doxygen is installed as part of the conda environment 
- The note in [3] be changed to acknowledge that obsolete tag warnings will be observed from time to time and may generally be ignored

For the best reproducibility between build environments, we might want to constrain the Doxygen version more tightly, specifying the same version in both the circleci config.yml and the conda environment.yml. I do understand that a) version constraints require downstream maintenance and b) this may be a minor issue within a larger conversation on reproducibility.

Why I think constraining the Doxygen version might be worth the maintenance: 
- Newcomers are often advised to pick out a documentation issue to work on as a first problem. They are likely to notice and be concerned about build warnings.
- Some parts of the documentation toolchain have been reported to be relatively brittle [4]. Constraining versions might keep newcomers and other humans from running into known problems that have been eliminated by how the circleci build is currently configured.

[1] CircleCI Build log:  https://app.circleci.com/pipelines/github/numpy/numpy/29054/workflows/1ec6c92c-a173-4a73-9080-2cf6f8b220d8/jobs/42939?invite=true#step-107-1162_27
[2] Condaforge download page for Doxygen: https://anaconda.org/conda-forge/doxygen
[3] How to build the docs: https://numpy.org/doc/stable/dev/howto_build_docs.html
[4] Doxygen download page: https://www.doxygen.nl/download.html
[5] Documentation tool chain 'fragile': https://github.com/numpy/numpy/issues/27377#issuecomment-2376281668

**Comment by ngoldbaum** [comment]
If constraining the doxygen version fixes warning I think it's fine to do that. However, the doxygen stuff really just needs someone to push it forward and make sure it gets used and updated regularly. Something like a rule in the CI that checks new C code for doxygen docstrings. It would also go a very long way to make sure the sphinx docs for the C code somehow is generated from the doxygen docstrings.

I don't want to go so far as to say we should completely rip out doxygen, since I think it would be a good improvement in the long run, but I also think having two sources of truth (the sphinx docs and the docstrings) is probably worse than not having doxygen running.

**Comment by mattip** [comment]
One problem we will have to deal with if we wish to keep c -> doxygen -> xml -> breathe -> rst -> sphinx -> html is that breathe is not really being supported and is starting to show some bitrot. There does not seem to be a good alternative.

**Comment by eagunn** [comment]
@mattip -- I fear that I have been not-understanding something you and others have said to me repeatedly because it simply didn't make sense to me. If so, I apologize for having a thick skull.

Are you saying that our doc build process:
- Requires and installs doxygen and breathe
- Runs doxygen over the .c/.cpp files to generate the XML output found in numpy\doc\build\doxygen\xml
- Does (or does not, I'm still not clear on this) run Breathe to convert that XML to RST
- In any case, abandons the output from .c/.cpp and does not feed it into the rest of the API documentation that is generated from Python code.

With the effect that:
- Any of the pages of API documentation found at or below file://wsl.localhost/Ubuntu/home/eagunn/openSource/numpy/doc/build/html/reference/c-api/index.html were generated straight from .rst input and were not generated from code and
- All of the Doxygen warning and version issues that I've been chewing on the edges of with my last couple of new issues are simply noise from a currently unused toolset and dead-end in the build process.

If so 
```
<humor type="self deprecating"> 
I have clearly failed the first test at the top of the c-api array page. file://wsl.localhost/Ubuntu/home/eagunn/openSource/numpy/doc/build/html/reference/c-api/array.html 
Sigh. 
</humor>
```

**Comment by mattip** [comment]
Yes, that is a very good approximation. The only such documentation that might be actually used is from the page describing how doxygen may be used for documentation, i.e. https://numpy.org/doc/stable/dev/howto-docs.html#documenting-c-c-code. Note how the "And here is how it is rendered:" is broken. It would be nice to fix or remove these.

**Comment by eagunn** [comment]
I just tried building the docs with and without the Doxygen step.

What's interesting is that the main difference in output that I've spotted so far is, with Doxygen, my local build renders the doxygen insertions on that page just fine.
![image](https://github.com/user-attachments/assets/34a0a7bb-faf0-491d-9514-8fd48733b62b)

It is only if I do not run Doxygen at all that I get the same error message as on the stable/deployed page:
![image](https://github.com/user-attachments/assets/1a37f5b8-1c94-4fd0-8bf1-1f5febb81e9f)

I'll see if I can figure out what is up with the official build. 

Meantime, if we decided to simply pull the Doxygen step out of the doc build process, then the whole section of that page
https://numpy.org/doc/stable/dev/howto-docs.html#documenting-c-c-code
that shows errors would go away. Now it not only shows errors but documents a process that is not working.

**Comment by eagunn** [comment]
I'm not seeing an obvious reason why the examples on that page are not rendering in the Circle build. The XML is being generated by Doxygen and I'm not seeing any messages about the Breathe extension being skipped or failing during the Sphinx step. 
I did look through the last few doc versions, though and can see that the documenting-c-c-code section was fine up through version 1.26 and then started to have the failed-render/warnings with the 2.0 build.

**Comment by eagunn** [comment]
I've looked at the Breathe project and it's a pretty discouraging picture. Their core maintainers don't seem to have energy or bandwidth right now even to merge pull requests that seem fairly complete. That must be very frustrating for them.

@mattip, Do you think the community/maintainers would be open to an issue proposing that we simply pull the Doxygen and Breathe steps out of the builds, pull that documenting-c-c-code section out of the documentation, and have a cleaner, less confusing doc build while we wait to see if Breathe can revive? 

I'd be willing to do the work to de-cruft the builds.

**Comment by mattip** [comment]
There were some commits to breathe `main` over the last few months. Does the current `HEAD` work?

**Comment by mattip** [comment]
`pip install git+https://github.com/breathe-doc/breathe.git@62695c7e3eaf1190f7bb02ba7b396daf78da66f4`

**Comment by mattip** [comment]
Breathe has [released a new version](https://github.com/breathe-doc/breathe/releases/tag/v4.36.0)

**Comment by InessaPawson** [comment]
Part of the issue has been addressed: the Doxyfile has been updated, its header now reads 1.13.2, matching the Doxygen version the build environment uses.

The rendering issue in [dev/howto-docs.html#documenting-c-c-code](https://numpy.org/devdocs/dev/howto-docs.html#documenting-c-c-code) mentioned in https://github.com/numpy/numpy/issues/27610#issuecomment-2427360288 has been fixed.

Better documenting the Doxygen setup would still be a valuable contribution.
